node/src/database/test_utils/transaction_wrapper_mock.rs

```diff
@@ -0,0 +1,394 @@
+
+// The previous paragraph clears up that we do not need to care about a mechanism to deliver errors
+// on the return of the 'prepare' method. The place worth our interest, though, is the produced
+// Statement coming out from this method. Even though looking perhaps a little nonsensical at first,
+// the 'prepare' method can have a strong impact on the result returned from the next, upstream
+// function call, to happen on top of this Statement, and which, luckily for us (despite certain
+// difficulties), can be indirectly shaped this way towards our requirements.
+// This distantly prepared Statement can then cause a certainly useful error, easing our test
+// writing. For some examples let's consider the following methods 'execute', 'query_row' or
+// 'query_map'.
+
+#[derive(Debug)]
+struct SetupForStubbed<S> {
+    stubbed_calls_conn: Box<dyn ConnectionWrapper>,
+    stubbed_calls_optional_statements_literals: Vec<S>,
+}
+
+#[derive(Debug)]
+struct SetupForBoth {
+    prod_code_calls_conn_used_for_both: bool,
+    prod_code_calls_conn: Box<dyn ConnectionWrapper>,
+    prod_code_calls_transaction_opt: Option<Box<dyn TransactionWrapper>>,
+    requested_preceding_prod_code_calls: usize,
+    stubbed: SetupForStubbed<Option<String>>,
+}
+
+impl SetupForBoth {
+    fn commit_prod_calls(&mut self) -> Result<(), Error> {
+        let txn_opt = self.prod_code_calls_transaction_opt.take();
+        let mut txn = txn_opt
+            .expect("Error: missing transaction in setup for both prod code and stubbed calls");
+        txn.commit()
+    }
+}
+
+impl Drop for SetupForBoth {
+    fn drop(&mut self) {
+        // Making sure that the referenced transaction will deconstruct
+        // before the connection it was pointing to
+        drop(self.prod_code_calls_transaction_opt.take())
+    }
+}
+
+#[derive(Debug)]
+pub struct PrepareMethodResults {
+    calls_counter: RefCell<usize>,
+    setup: Either<SetupForBoth, SetupForStubbed<String>>,
+}
+
+impl PrepareMethodResults {
+    pub fn new_with_both_prod_code_and_stubbed_calls(
+        prod_code_calls_conn: Box<dyn ConnectionWrapper>,
+        stubbed_calls_conn: Box<dyn ConnectionWrapper>,
+    ) -> Self {
+        let setup = {
+            let ptr = Box::into_raw(prod_code_calls_conn);
+            let conn = unsafe { Box::from_raw(ptr) };
+
+            let mut setup = SetupForBoth {
+                prod_code_calls_conn_used_for_both: false,
+                prod_code_calls_conn: conn,
+                prod_code_calls_transaction_opt: None,
+                requested_preceding_prod_code_calls: 0,
+                stubbed: SetupForStubbed {
+                    stubbed_calls_conn,
+                    stubbed_calls_optional_statements_literals: vec![],
+                },
+            };
+
+            let conn = unsafe { ptr.as_mut().unwrap() };
+            let txn = conn.transaction().unwrap();
+
+            setup.prod_code_calls_transaction_opt = Some(txn);
```

Comment:
Nope. The whole point is I need to store two things that reference to each other inside two fields of a struct. That's kind of impossible thing to do unless you invalidate the reference somehow, that's why I need the raw pointer and still I have to be careful about implementing the rest. 

This is also where all those "drops" etc come from, preventing OS errors. I've rewritten it hugely so that it's is clearer and safer, and made comments which believe are understandable.